src/util/expr_iterator.h

```diff
@@ -0,0 +1,242 @@
+/*******************************************************************\
+
+ Module: exprt iterator module
+
+ Author: DiffBlue Limited. All rights reserved.
+
+\*******************************************************************/
+
+#ifndef CPROVER_UTIL_EXPR_ITERATOR_H
+#define CPROVER_UTIL_EXPR_ITERATOR_H
+
+#include <iterator>
+#include <functional>
+#include <set>
+#include <algorithm>
+#include "expr.h"
+#include "invariant.h"
+
+// Forward declarations - table of contents
+
+/// \file Forward depth-first search iterators
+/// These iterators' copy operations are expensive, so use auto&, and avoid
+/// std::next(), std::prev() etc.
+///
+/// Incrementable only by calling ++it (to avoid accidental copying, as
```

Comment:
> STL algorithms already support pre-increment version
I'm not convinced, and [this person on stack overflow agrees](https://stackoverflow.com/a/23305753). I think programmers are generally aware that post-increment is likely to cause unnecessary copies (this often comes up in coding guidelines) and write code accordingly. However, disallowing post-increment is surprising (if in doubt, "do as the `int`s do") *and violates the requirements of the `forward_iterator` model*.